File: RemoveK.java

class RemoveK {
    // Time Complexity : O(N) where N is length of num array
    // Space Complexity : O(N) size of the stack
    // Did this code successfully run on Leetcode : Yes
    // Any problem you faced while coding this :No


    /**
     * Approach:
     * 1. Add digits to stack. If the next digit is less than top element of stack, pop the element.
     * 2. At the end reverse the elements of stack and return the output
     */

    public String removeKdigits(String num, int k) {
        
        if(num == null || num.length() == 0 || k < 1){
            return num;
        }
        
        if(num.length() <= k){
            return "0";
        }
        
        Stack<Integer> stack = new Stack();
        int count = 0;
        int i = 0;
        StringBuilder sb = new StringBuilder();
        
        while(i < num.length()){
            
            while(count < k && !stack.isEmpty() && num.charAt(i) < num.charAt(stack.peek())){
                stack.pop();
                count++;
            }
            stack.push(i);
            i++;
            
        }
        
        while(count < k){
            stack.pop();
            count++;
        }
        
        while(!stack.isEmpty()) {
            sb.insert(0, num.charAt(stack.pop()));
        }
        
        while(sb.length() > 1 && sb.charAt(0) == '0'){
            sb.deleteCharAt(0);
        }
        
        return sb.length() == 0 ? "0" : sb.toString();
        
    }
}

File: remove_k.py
"""
# Time complexity - O(n) 
# Space Complexity: O(n)
# Did this code successfully run on LeetCode?: Yes
# Problems faced while coding this:None
Approach: Using stack:
    append stack unless the values are smaller than the current element
    if current element is smaller than top of the stack, then remove that number and decrement k value by 1
    pop elements until k becomes 0 (in case after visiting each element the length of the stack is larger than req)
    lastly take care of leading 0's
"""

class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        numStack = []
        
        for digit in num:
            #print(numStack)

            while k and numStack and numStack[-1] > digit:
                numStack.pop()
                k -= 1
            numStack.append(digit)
        
        final = numStack[:-k] if k else numStack
        
        # leading zeros
        return "".join(final).lstrip('0') or "0"